XMonad/Hooks/DynamicIcons.hs

```diff
@@ -95,29 +95,40 @@ dynamicIconsPP ic pp = getWorkspaceIcons ic <&> \ren -> pp{ ppRename = ppRename
 -- | Returns a function for 'ppRename' that augments workspaces with icons
 -- according to the provided 'IconConfig'.
 getWorkspaceIcons :: IconConfig -> X (String -> WindowSpace -> String)
-getWorkspaceIcons IconConfig{..} = fmt <$> getWorkspaceIcons' iconConfigIcons
+getWorkspaceIcons IconConfig{..} = fmt <$> iconFilterFunction iconConfigIcons
   where
     fmt icons s w = iconConfigFmt s (M.findWithDefault [] (S.tag w) icons)
 
+-- | Only use the focused window for each workspace to find icon
+getFocusedIcon :: Query [String] -> X (M.Map WorkspaceId [String])
+getFocusedIcon q = do
+    ws <- gets (S.workspaces . windowset)
+    is <- for ws $  fmap (concat . maybeToList) . traverse (runQuery q . S.focus)  . S.stack
+    pure $ M.fromList (zip (map S.tag ws) is)
+
+-- | Use all icons for each workspace
 getWorkspaceIcons' :: Query [String] -> X (M.Map WorkspaceId [String])
 getWorkspaceIcons' q = do
     ws <- gets (S.workspaces . windowset)
-    is <- for ws $ foldMap (runQuery q) . S.integrate' . S.stack
+    is <- flip foldMap ws $ traverse (runQuery q .  S.focus) . maybeToList . S.stack
     pure $ M.fromList (zip (map S.tag ws) is)
 
 
 -- | Datatype for expanded 'Icon' configurations
 data IconConfig = IconConfig
-    { iconConfigIcons :: Query [String]
+    { iconConfigIcons    :: Query [String]
       -- ^ What icons to use for each window.
-    , iconConfigFmt :: WorkspaceId -> [String] -> String
+    , iconConfigFmt      :: WorkspaceId -> [String] -> String
       -- ^ How to format the result, see 'iconsFmtReplace', 'iconsFmtAppend'.
+    , iconFilterFunction :: Query [String] -> X (M.Map WorkspaceId [String])
```

Comment:
This type seems quite a bit more complex than what its description says about it. It would probably be fine if the type was opaque and users were only meant to choose between `getFocusedIcon` and `getWorkspaceIcons'` (the prime being one of the hints that exporting this one might not be the best of ideas), but we do expose the type and we do expect people to use it.

Can we perhaps change this to `Maybe (Stack Window) → X [Window]` or something like that, with `pure . S.integrate'` being the default and `getFocusedIcon` being just `pure . maybeToList . fmap S.focus`?

Reply:
I made it something along those lines, including getAllIcons